Sources/Plasma/NucleusLib/pnAsyncCoreExe/pnAceSocket.cpp

```diff
@@ -0,0 +1,553 @@
+    }
+
+    tcp::endpoint remoteEndpoint = sock->fSock.remote_endpoint(err);
+    if (err) {
+        LogMsg(kLogError, "Failed to get remote socket endpoint: {}", err.message());
+    } else if (remoteEndpoint.address().is_v4()) {
+        remoteAddr->SetHost(remoteEndpoint.address().to_v4().to_bytes());
+        remoteAddr->SetPort(remoteEndpoint.port());
+    } else {
+        remoteAddr->Clear();
+        LogMsg(kLogError, "Remote socket endpoint is not IPv4");
+    }
+}
+
+static bool SocketInitConnect(ConnectOperation* op)
+{
+    // This steals ownership of op->fSock
+    auto sock = std::make_unique<AsyncSocketStruct>(op);
+    
+    sock->initTimeMs    = TimeGetMs();
+
+    asio::error_code err;
+    sock->fSock.non_blocking(true, err);
+    if (err)
+        LogMsg(kLogError, "Failed to set socket to non-blocking: {}", err.message());
+
+    // Send initial data
+    if (!op->fConnectBuffer.empty()) {
+        if (!AsyncSocketSend(sock.get(), op->fConnectBuffer.data(), op->fConnectBuffer.size()))
+            return false;
+    }
+
+    if (!IS_TEXT_CONNTYPE(op->fConnectionType)
+        && op->fConnectBuffer.size() < sizeof(AsyncSocketConnectPacket))
+    {
+        return false;
+    }
+
+    // perform callback notification
+    AsyncNotifySocketConnect notify;
+    SocketGetAddresses(sock.get(), &notify.localAddr, &notify.remoteAddr);
+    notify.param        = sock->fParam;
+    notify.asyncId      = nullptr;
+    notify.connType     = sock->fConnectionType;
+    if (!sock->fNotifyProc(sock.get(), kNotifySocketConnectSuccess, &notify, &sock->fUserState))
+        return false;
+
+    // start reading from the socket
+    SocketStartAsyncRead(sock.release());
+    return true;
+}
+
+void AsyncSocketConnect(AsyncCancelId* cancelId, const plNetAddress& netAddr,
+                        FAsyncNotifySocketProc notifyProc, void* param,
+                        const void* sendData, unsigned sendBytes)
+{
+    ASSERT(notifyProc);
+    ASSERT(s_ioPool);
+
+    PerfAddCounter(kAsyncPerfSocketConnAttemptsOutCurr, 1);
+    PerfAddCounter(kAsyncPerfSocketConnAttemptsOutTotal, 1);
+
+    auto op = new ConnectOperation(s_ioPool->fContext);
+    op->fRemoteAddr = netAddr;
+    op->fNotifyProc = std::move(notifyProc);
+    op->fParam = param;
+    if (sendBytes) {
+        auto sendBuffer = reinterpret_cast<const uint8_t*>(sendData);
+        op->fConnectBuffer.assign(sendBuffer, sendBuffer + sendBytes);
+        op->fConnectionType = sendBuffer[0];
+    } else {
+        op->fConnectionType = kConnTypeNil;
+    }
+
+    hsLockGuard(s_connectCrit);
+
+    if (cancelId) {
+        // get cancel id; we can avoid checking for zero by always using an odd number
+        ASSERT(s_nextConnectCancelId & 1);
+        s_nextConnectCancelId += 2;
+
+        *cancelId = op->fCancelId = (AsyncCancelId)s_nextConnectCancelId;
+    }
+
+    s_connectList.emplace_back(op);
+
+    tcp::endpoint remoteAddr(asio::ip::address_v4(netAddr.GetHostBytes()),

```

Comment:
This could probably just use the old `GetHost()` method with a `reinterpret_cast<uint8_t*>(&host);` instead of requiring the addition of an entirely new method in `plNetAddress`.

Reply:
I generally agree with this comment, but I'll let this sit for a bit in case anyone has any other feedback. Adding a function where one wasn't seems more intentional and I didn't know if there was a deeper architectural reasoning here.

Reply:
Looking through this one again. Here's what I'll propose: I think this change should be implemented, but that GetHostBytes should stay. plNetAddress is a nice utility class and GetHostBytes could be useful for debugging. But the current implementation does a bit of unnecessary work.